src/scenario/scenario.c

```diff
@@ -539,21 +562,43 @@ void scenario_objectives_from_buffer(buffer *buf, int version, scenario_win_crit
     } else {
         buffer_set(buf, 1640);
     }
-    win_criteria->culture.goal = buffer_read_i32(buf);
-    win_criteria->prosperity.goal = buffer_read_i32(buf);
-    win_criteria->peace.goal = buffer_read_i32(buf);
-    win_criteria->favor.goal = buffer_read_i32(buf);
-    win_criteria->culture.enabled = buffer_read_u8(buf);
-    win_criteria->prosperity.enabled = buffer_read_u8(buf);
-    win_criteria->peace.enabled = buffer_read_u8(buf);
-    win_criteria->favor.enabled = buffer_read_u8(buf);
-    win_criteria->time_limit.enabled = buffer_read_i32(buf);
-    win_criteria->time_limit.years = buffer_read_i32(buf);
-    win_criteria->survival_time.enabled = buffer_read_i32(buf);
-    win_criteria->survival_time.years = buffer_read_i32(buf);
+    scenario_criteria_clear();
+    int culture_goal = buffer_read_i32(buf);
+    int prosperity_goal = buffer_read_i32(buf);
+    int peace_goal = buffer_read_i32(buf);
+    int favor_goal = buffer_read_i32(buf);
+    int culture_enabled = buffer_read_u8(buf);
+    int prosperity_enabled = buffer_read_u8(buf);
+    int peace_enabled = buffer_read_u8(buf);
+    int favor_enabled = buffer_read_u8(buf);
+    int time_limit_enabled = buffer_read_i32(buf);
+    int time_limit = buffer_read_i32(buf);
+    int survival_time_enabled = buffer_read_i32(buf);
+    int survival_time_years = buffer_read_i32(buf);
+    if (culture_enabled) {
+        scenario_criteria_try_add_or_update_ptr(win_criteria, WIN_CRITERIA_CULTURE, culture_goal, 0);
```

Comment:
I might be reading this wrong, but I would not save the data onto the existing scenario object.
(I'm now talking beyond the standard win criteria that was already in there, since that is also read on loading and displaying scenarios.)

For the new goals array, I would save it as an entirely new file_piece, since you then do not have any of the headaches of backwards compatibility, and can use dynamic sizing for saving / loading the array itself.

Then instead of loading goals from the scenario you can read and display the goals array. For old scenarios / savegames, you can read these values from the scenario piece, and 'generate' standard goals in the array (after which your normal code handles them as if they were there the whole time.)

Reply:
My goal with this first PR was not to touch serialization at all. Right now only the existing win criteria (culture, prosperity, peace, favor, population, time limit and survival time) are saved in the .sav file or .map file, and I am reading/writing them in exactly the same order they are in now. Future work will need to change serialization to support a wider array of criteria.

That said, if the number of possible criteria is relatively small, and if the criteria objects themselves are not too bloated, we might be able to keep them in the existing file segment. We have to figure out how complicated the criteria are, before we can figure out where to find the space to store them.

Reply:
That makes sense.

You might get the serialization for free if you piggyback off the serialization that I am doing for conditions.
My plan there is that conditions and action can serialize on their own (apart from events) since they might be linked to other things in the future. such as win criteria or requests.